File: src/xianyu_cli/utils/cookie.py
"""Browser cookie extraction for Goofish/XianYu domains.

Uses browser-cookie3 to extract cookies from installed browsers.
Supports Chrome, Firefox, Edge, Safari, Brave, Opera, etc.
"""

from __future__ import annotations

import logging
from http.cookiejar import CookieJar
from typing import Any

from xianyu_cli.utils._common import COOKIE_DOMAINS

logger = logging.getLogger(__name__)

# Browser extraction functions from browser-cookie3
_BROWSER_LOADERS: dict[str, str] = {
    "chrome": "chrome",
    "firefox": "firefox",
    "edge": "edge",
    "safari": "safari",
    "brave": "brave",
    "opera": "opera",
    "vivaldi": "vivaldi",
}


def extract_browser_cookies(
    browser: str | None = None,
    domains: list[str] | None = None,
) -> dict[str, str] | None:
    """Extract cookies from a browser for Goofish domains.

    Args:
        browser: Specific browser to extract from. None tries all browsers.
        domains: Cookie domains to filter. Defaults to Goofish/Taobao domains.

    Returns:
        Dict of cookie name→value, or None if extraction failed.
    """
    try:
        import browser_cookie3
    except ImportError:
        logger.warning("browser-cookie3 not installed")
        return None

    domains = domains or COOKIE_DOMAINS
    browsers_to_try: list[str] = [browser] if browser else list(_BROWSER_LOADERS.keys())

    for br_name in browsers_to_try:
        loader_name = _BROWSER_LOADERS.get(br_name)
        if not loader_name:
            continue
        loader_fn = getattr(browser_cookie3, loader_name, None)
        if not loader_fn:
            continue

        try:
            cj: CookieJar = loader_fn(domain_name="")
            cookies = _filter_cookies(cj, domains)
            if cookies and _has_required_cookies(cookies):
                logger.info("Extracted cookies from %s", br_name)
                return cookies
        except Exception:
            logger.debug("Failed to extract from %s", br_name, exc_info=True)
            continue

    return None


def _filter_cookies(cj: CookieJar, domains: list[str]) -> dict[str, str]:
    """Filter cookie jar to only include cookies for specified domains."""
    result: dict[str, str] = {}
    for cookie in cj:
        if any(cookie.domain.endswith(d.lstrip(".")) for d in domains):
            if cookie.value:
                result[cookie.name] = cookie.value
    return result


def _has_required_cookies(cookies: dict[str, str]) -> bool:
    """Check if the extracted cookies represent a truly authenticated session.

    Requires both _m_h5_tk (for sign) AND unb (user ID, only set after login).
    Without unb, the cookies are just visitor cookies and API calls will fail.
    """
    return "_m_h5_tk" in cookies and "unb" in cookies


def parse_cookie_string(cookie_str: str) -> dict[str, str]:
    """Parse a raw Cookie header string into a dict."""
    cookies: dict[str, str] = {}
    for part in cookie_str.split(";"):
        part = part.strip()
        if "=" in part:
            key, _, value = part.partition("=")
            cookies[key.strip()] = value.strip()
    return cookies

File: golang/utils/cookie.go
package utils

import (
	"strings"
)

// ParseCookieString parses a raw Cookie header string into a map.
func ParseCookieString(cookieStr string) map[string]string {
	cookies := make(map[string]string)
	for _, part := range strings.Split(cookieStr, ";") {
		part = strings.TrimSpace(part)
		if idx := strings.Index(part, "="); idx > 0 {
			key := strings.TrimSpace(part[:idx])
			value := strings.TrimSpace(part[idx+1:])
			cookies[key] = value
		}
	}
	return cookies
}

// HasRequiredCookies checks if cookies represent an authenticated session.
func HasRequiredCookies(cookies map[string]string) bool {
	_, hasTK := cookies["_m_h5_tk"]
	_, hasUNB := cookies["unb"]
	return hasTK && hasUNB
}

// BuildCookieHeader builds a Cookie header string from a map.
func BuildCookieHeader(cookies map[string]string) string {
	parts := make([]string, 0, len(cookies))
	for k, v := range cookies {
		parts = append(parts, k+"="+v)
	}
	return strings.Join(parts, "; ")
}

// ExtractBrowserCookies attempts to extract cookies from installed browsers.
// Note: Go doesn't have a direct equivalent of browser-cookie3.
// This is a stub that returns nil - users should use --cookie flag instead.
func ExtractBrowserCookies(browser string) map[string]string {
	// Browser cookie extraction requires platform-specific code
	// (reading Chrome SQLite DB, decrypting with Keychain on macOS, etc.)
	// For now, recommend using --cookie flag to provide cookies manually.
	return nil
}
